beetsplug/mbweb.py

```diff
@@ -0,0 +1,109 @@
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""Runs a local HTTP server for MusicBrainz web tagger and handles
+users' requests.
+"""
+
+from __future__ import division, absolute_import, print_function
+
+import socket
+import webbrowser
+from io import BytesIO
+from six.moves.urllib_parse import urlparse, urlencode, parse_qs
+from six.moves.BaseHTTPServer import BaseHTTPRequestHandler
+from beets import autotag
+from beets.plugins import BeetsPlugin
+from beets.ui.commands import PromptChoice
+from beets import ui
+
+PORT = 8000
+
+
+# Simple HTTP parser
+class HTTPRequest(BaseHTTPRequestHandler):
+    def __init__(self, request_text):
+        self.rfile = BytesIO(request_text)
+        self.raw_requestline = self.rfile.readline()
+        self.parse_request()
+
+
+# Get 'id' parameter from URI
+def parse(data):
+    r = HTTPRequest(data)
+    parsed = urlparse(r.path)
+    return parse_qs(parsed.query)['id'][0]
+
+
+class Server():
+    def __init__(self):
+        self.host = '127.0.0.1'
+        self.port = PORT
+        try:  # Start TCP socket, catch soket.error
+            self.run_server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            self.run_server.setsockopt(socket.SOL_SOCKET,
+                                       socket.SO_REUSEADDR, 1)
+            self.run_server.bind((self.host, self.port))
+        except socket.error as error_msg:
+            print("Error occurred: {0}".format(error_msg))
+
+    def listen(self, size=1024):
+        self.run_server.listen(1)
+        while True:
+            con, addr = self.run_server.accept()
+            while True:
+                data = con.recv(size)
+                con.send(data)
+                if not data:
+                    break
+                return parse(data)
+
+
+class MBWeb(BeetsPlugin):
+    def __init__(self):
+        super(MBWeb, self).__init__()
+        self.port = PORT
+        self.register_listener('before_choose_candidate', self.prompt)
+        self.server = Server()
+
+    def prompt(self, session, task):
+        return [PromptChoice('l', 'Lookup', self.choice)]
+
+    def choice(self, session, task):
+        search_ids = []
+        query = {'tport': self.port,
+                 'artist': None,
+                 'track': None,
+                 'release': None,
+                 }
+        ui.print_("Choose your track and click 'tagger' button to add.")
+        if task.is_album:
+            for item in task.items:
+                query['artist'] = item.albumartist
+                query['track'] = item.title
+                query['release'] = item.album
+                url = 'http://musicbrainz.org/taglookup?{0}'\
+                    .format(urlencode(query))
+                webbrowser.open(url)
+                id_choice = self.server.listen()
+                search_ids.append(id_choice)
```

Comment:
Hmm; looks like this opens the browser for every track on the album? Maybe we just want to search by album?